modules/standard/Random.chpl

```diff
@@ -973,21 +977,56 @@ module Random {
                 element type, with no negative values and at least one non-zero
                 value. The domain must be equal to that of ``arr.domain``.
 
-     :return: An element chosen from ``arr`` is ``size == 1``, or an array of
+     :return: An element chosen from ``arr`` if ``size == 1``, or an array of
               element chosen from ``arr`` if ``size > 1`` or ``size`` is a
               domain.
 
      :throws IllegalArgumentError: if ``arr.size == 0``,
                                    if ``arr`` contains a negative value,
-                                   if ``arr`` has no non-zero values.,
-                                   if ``arr.domain != prob.domain``,
+                                   if ``arr`` has no non-zero values,
                                    if ``size < 1 || size.size < 1``,
                                    if ``replace=false`` and ``size > arr.size || size.size > arr.size``
      */
-      proc choice(arr: [], size:?sizeType=none, replace=true, prob:?probType=none)
+      proc choice(arr: [?Dom], size:?sizeType=none, replace=true, prob:?probType=none)
         throws
       {
-        return _choice(this, arr, size=size, replace=replace, prob=prob);
+        return _choice(this, Dom, size=size, replace=replace, prob=prob);
```

Comment:
`_choice(Dom,...)` returns an index or indices from the `Dom`, right?  This function `choice(arr: [?Dom],...)` should return elements of the array, not indices of its domain.

Reply:
@cassella thanks for reviewing. The `_choice(Dom, ...)` function do not return indices, it returns elements from `Dom` which are indices of the passed array/range.

Reply:
Right.  But the `choice(arr: [?Dom],)` function should return elements of `arr`, not elements of `Dom`.

What does `stream.choice(["a"])` return?

Reply:
> What does `stream.choice(["a"])` return?

@cassella this will return `0`.

> But the `choice(arr: [?Dom],)` function should return elements of `arr`, not elements of `Dom`.

I think this function was expected to return indices of `arr` in the new implementation, which are elements of `Dom` of that array. In the `_choice()` function we are randomly generating an index for the `Dom` and then mapping it to an element of the `Dom` which in turn is an indice of the `arr`.

Reply:
The existing function `choice(arr: [?Dom],...)` returns an array element.

I think the request in #14865 is to add two new versions of
`choice()`: one taking a domain and another taking a range, that
return indices.  I believe it was requesting those in addition to the
existing behavior.

That is, ending up with three versions:

`choice(arr: [?Dom], ...)` - returns an array element
`choice(D: domain, ...)` - returns an index from the domain
`choice(r: range, ...)` - returns an index from the range.

And so

```chapel
var A : [D] string;
stream.choice(A); // returns a string
stream.choice(A.domain); // returns an index in D
```

I don't see anything in the request suggesting to change the behavior
of existing calls to `choice(arr[])`.


I think this is why @ben-albrecht was expecting that `var counts = new
map(eltType, int)` in the array test case to continue counting `eltType`s.

Reply:
@cassella thanks for explaining, I think I misinterpreted @ben-albrecht's [comment](https://github.com/chapel-lang/chapel/issues/14865#issuecomment-585764974). I thought we do not want to keep the old functionality, as random elements of an array can be obtained if we have random indices. Currently, we have two overloads some for `arr` and one for `rng`,
and `_choice` accepts domain and return random elements from the domain.

Now to return elements in case of an array, the array should also be passed to the `_choice` function. But in case of domain and range where we have to return indices there is no need of an array. Should I add an optional array arg to `_choice` just like the `prob` arg? What do you suggest?

Reply:
That's just my take; you can wait for Ben or Brad to clarify.

I don't think you'd have to make that drastic a change -- you're almost there already:

```chapel
proc choice(arr: [?Dom], ...) {
    var idx = _choice(this, Dom, ...);
    return arr[idx];
}
```

Reply:
> ```chapel
> proc choice(arr: [?Dom], ...) {
>     var idx = _choice(this, Dom, ...);
>     return arr[idx];
> }
> ```

This seems to be a good approach. Thanks!

> That's just my take; you can wait for Ben or Brad to clarify.

Yes, let's ask @ben-albrecht and @bradcray what they think?

Reply:
@cassella's interpretation is correct:

That is, ending up with three versions:

> `choice(arr: [?Dom], ...)` - returns an array element
> `choice(D: domain, ...)` - returns an index from the domain
> `choice(r: range, ...)` - returns an index from the range.

Reply:
> That is, we should end up with three overloads:

I will make the necessary changes. Thanks :-)

Reply:
OK, feel free to ping me when ready for review again.

Reply:
> OK, feel free to ping me when ready for review again.

@ben-albrecht I have made the changes. Added overload for domain and modified array overload to return elements like before. I added/updated the tests accordingly ;-)